benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/AbstractBenchmarkRunner.java

```diff
@@ -0,0 +1,762 @@
+            stockLevel(statement);
+            statistics.incStockLevel();
+          } else {
+            LOG.info("No transaction");
+          }
+          if (Thread.interrupted()) {
+            break;
+          }
+        } catch (Throwable exception) {
+          if ((exception instanceof PSQLException psqlException)
+              && psqlException.getServerErrorMessage() != null
+              && Objects.equals(
+                  psqlException.getServerErrorMessage().getSQLState(),
+                  PSQLState.SERIALIZATION_FAILURE.getState())
+              && Objects.requireNonNull(psqlException.getServerErrorMessage().getMessage())
+                  .contains("concurrent modification")) {
+            LOG.debug("Transaction aborted by Cloud Spanner");
+            statistics.incAborted();
+          } else {
+            LOG.warn("Transaction failed", exception);
+            statistics.incFailed();
+          }
+          execute(statement, "rollback");
+        }
+      }
+    }
+  }
+
+  private long getOtherWarehouseId(long currentId) {
+    if (tpccConfiguration.getWarehouses() == 1) {
+      return currentId;
+    }
+    while (true) {
+      long otherId = Long.reverse(random.nextInt(tpccConfiguration.getWarehouses()));
+      if (otherId != currentId) {
+        return otherId;
+      }
+    }
+  }
+
+  private void newOrder(Statement statement) throws SQLException {
+    LOG.debug("Executing new_order");
+
+    long warehouseId = Long.reverse(random.nextInt(tpccConfiguration.getWarehouses()));
+    long districtId = Long.reverse(random.nextInt(tpccConfiguration.getDistrictsPerWarehouse()));
+    long customerId = Long.reverse(random.nextInt(tpccConfiguration.getCustomersPerDistrict()));
+
+    int orderLineCount = random.nextInt(5, 16);
+    long[] itemIds = new long[orderLineCount];
+    long[] supplyWarehouses = new long[orderLineCount];
+    int[] quantities = new int[orderLineCount];
+    int rollback = random.nextInt(100);
+    int allLocal = 1;
+
+    for (int line = 0; line < orderLineCount; line++) {
+      if (rollback == 1 && line == orderLineCount - 1) {
+        itemIds[line] = Long.reverse(Long.MAX_VALUE);
+      } else {
+        itemIds[line] = Long.reverse(random.nextInt(tpccConfiguration.getItemCount()));
+      }
+      if (random.nextInt(100) == 50) {
+        supplyWarehouses[line] = getOtherWarehouseId(warehouseId);
+        allLocal = 0;
+      } else {
+        supplyWarehouses[line] = warehouseId;
+      }
+      quantities[line] = random.nextInt(1, 10);
+    }
+
+    Object[] row;
+    execute(statement, "begin transaction");
+    row =
+        queryRow(
+            statement,
+            (tpccConfiguration.isLockScannedRanges() ? "/*@ lock_scanned_ranges=exclusive */" : "")
+                + "SELECT c_discount, c_last, c_credit, w_tax "
+                + "FROM customer c, warehouse w "
+                + "WHERE w.w_id = %d AND c.w_id = w.w_id AND c.d_id = %d AND c.c_id = %d",
```

Comment:
All these queries appear to use String.format(..) to add literals to the statements. The correct way to do that in JDBC would be to use a `java.sql.PreparedStatement` and add it as query parameters. This will otherwise cause Spanner to have to re-parse the query every time, as the literals change for each execution.

Reply:
Updated to use `java.sql.PreparedStatement`.